File: Data Structures & Algorithms/string-encode-and-decode/submission-0.java
class Solution {

    public String encode(List<String> strs) {
        StringBuilder res = new StringBuilder();
        for (String s : strs) {
            int len = s.length();
            res.append(len);
            res.append("#");
            res.append(s);
        }
        return res.toString();
    }

    public List<String> decode(String str) {
        List<String> res = new ArrayList<>();
        int i = 0;

        while (i < str.length()) {
            int j = i;

            while (str.charAt(j) != '#') {
                j++;
            }

            int length = Integer.parseInt(str.substring(i, j));
            int start = j + 1;
            res.add(str.substring(start, start + length));
            
            i = start + length;
        }
        return res;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-0.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        memo = {}
        def dfs(i, j):
            if i + j == len(s3):
                return (i == len(s1)) and (j == len(s2))
            
            if (i, j) in memo:
                return memo[(i, j)]
            
            res = False
            if i < len(s1) and s1[i] == s3[i + j]:
                res = dfs(i + 1, j)
            
            if j < len(s2) and s2[j] == s3[i + j]:
                res = dfs(i, j + 1)
            
            memo[(i, j)] = res
            return res

        return dfs(0, 0)
    
            